lib/logging.dart

```diff
@@ -127,7 +127,7 @@ class Logger {
   void clearListeners() {
     if (hierarchicalLoggingEnabled || parent == null) {
       if (_controller != null) {
-        _controller.close();
+        _controller!.close();
```

Comment:
This is a correct migration.  Even though we null-test `_controller` on the previous line, that null test doesn't promote, because `_controller` is a field, so it could be changed at any time by non-local code.

Reply:
In this particular case, `_controller` is a private field, and its not possible that local could have changed it (Dart is single threaded, and you cannot outside of the current library, override _controller, and I am assuming that can be proven?)

Reply:
> In this particular case, `_controller` is a private field, and its not possible that local could have changed it (Dart is single threaded, and you cannot outside of the current library, override _controller, and I am assuming that can be proven?)

Yes, it could certainly be proven in this case, especially because there is no code between the null check and the usage of the field.  And you're definitely not alone in wanting this--many others have asked for promotion to work on fields.

I spent quite a while digging into the idea when I was proposing the flow analysis rules for NNBD.  Here's roughly what I considered:
- We could promote private final fields that are not overridden in the current library.  That unfortunately wouldn't address this case because `_controller` isn't private (and it can't be due to the assignments in `clearListeners` and `_getStream`).
- We could promote private fields that are not overridden in the current library, provided that there is no code in between the test and the usage.  That would address this case, but it would lead to some counterintuitive behavior (for example, if you added an `assert` statement prior to the use of `_controller`, then promotion would stop working; or if you tried to use `_controller` twice, the second usage would fail).
- We could analyze any code between the test and the usage to see if assignments to the field are reachable from it.  This, unfortunately, is really expensive in the general case, so it's not something we could do in the front end without a big hit to compilation times, and fast compilation is really important.
- We could say that fields always promote, and if the compiler can't prove that the field isn't changed between the test and the usage, it just adds invisible null checks to restore soundness.  That has some negative consequences though.  First, part of the point of NNBD is to allow the user to improve the efficiency of their code by eliminating unnecessary null checks, so having the compiler insert invisible null checks seems counterproductive.  Second, if one of these silently inserted null checks fails, then it's a big source of potential user confusion.

In the end I couldn't find a way to make fields promotable that seemed sufficiently useful to justify the complexity.

Note that we're thinking of offering the user a quick fix that would transform this code into:
```dart
var _controller = this._controller;
if (_controller != null) {
  _controller.close();
}
```